posthog/rbac/user_access_control.py

```diff
@@ -444,6 +446,65 @@ def filter_queryset_by_access_level(self, queryset: QuerySet, include_all_if_adm
 
         return queryset
 
+    def filter_and_annotate_file_system_queryset(self, queryset: QuerySet["FileSystem"]) -> QuerySet["FileSystem"]:
+        """
+        Annotate each FileSystem with the effective_access_level and exclude items with
+        level = 'none' (unless the user is the creator or is staff/org-admin).
+        """
+
+        user = self._user
+        org_membership = self._organization_membership
+
+        # If staff or organization admin => skip extra filtering; user can see everything
+        if user.is_staff or (org_membership and org_membership.level >= OrganizationMembership.Level.ADMIN):
+            return queryset
+
+        # 1) We’ll map each `access_level` to an integer: none=0, viewer=1, editor=2 (default=2)
+        access_level_order_case = Case(
+            When(access_level="none", then=Value(0)),
+            When(access_level="viewer", then=Value(1)),
+            When(access_level="editor", then=Value(2)),
+            default=Value(2),  # default to 'editor'
+            output_field=IntegerField(),
+        )
+
+        # 2) Build a Subquery that finds the highest relevant access level row
+        #    matching (team_id, resource=FileSystem.type, resource_id=FileSystem.ref).
+        #    Must match at least one of these conditions:
+        #       - organization_member__user = self._user
+        #       - role__in = self._user_role_ids
+        #       - organization_member = None and role = None (global default)
+        highest_level_subquery = (
+            AccessControl.objects.filter(
+                team_id=OuterRef("team_id"),
+                resource=OuterRef("type"),
+                resource_id=OuterRef("ref"),
+            )
+            .filter(
+                Q(organization_member__user=user)
+                | Q(role__in=self._user_role_ids)
+                | Q(organization_member=None, role=None)
+            )
+            .annotate(level_value=access_level_order_case)
+            .order_by(-F("level_value"))
+            .values("level_value")[:1]
+        )
+
+        # 3) Annotate the queryset with 'effective_access_level'.
+        queryset = queryset.annotate(
+            effective_access_level=Coalesce(
```

Comment:
In the new access control system we aren't using `effective_access_level`, any uses of it are legacy. In this case, I see why you did it because it's not the exact resource but a pointer to it. Just wanted to call that out.